### DISCUSSION TITLE
Support for Apache-Ignite

### DISCUSSION THREAD

**Discussion by matsvitt** [apache/superset#29340] [Show and tell]
Hello everyone, 

if anyone is interested in accessing Apache Ignite datasets… this is now possible (however, early alpha - you’ve been warned).
Please include the latest version (which is 0.5.6 as of today) of the __sqlalchemy-ignitex__ project 
In your superset setup. It is installable via pipy via:

`pip install sqlalchemy-ignitex`

The code for this package is also on github (https://github.com/matsvitt/sqlalchemy-ignite/). It is a fork based on the original sqlalchemy-ignite project (kudos to predictiveworks). I just added a few things in order to make it work again.

Please let me know, if it is useful. I would be glad to support.

**Comment by fedyanabramovich** [comment]
Hey! Trying to connect my Ignite to superset.
Database itself works fine.
Added `sqlalchemy-ignitex` to `docker/requirements-local.txt` in the superset dist.
Looks like superset successfully installs your package but when I'm trying to connect my database to superset, I have an error.
Could you help me please? What I'm doing wrong?

<img width="810" alt="Screenshot 2025-01-24 at 16 41 17" src="https://github.com/user-attachments/assets/c89d739e-8621-46fe-b67e-7e1fdf4c9230" />
<img width="1728" alt="Screenshot 2025-01-24 at 17 25 47" src="https://github.com/user-attachments/assets/094ead06-7b27-4fd4-b19d-442f50a42c5f" />
<img width="1148" alt="Screenshot 2025-01-24 at 16 44 20" src="https://github.com/user-attachments/assets/cdf931ba-d1ae-45e0-a6c2-5af2776e94d7" />

**Comment by matsvitt** [comment]
Hey,

I thought for ages, I am the only one interested in connecting Superset to Ignite :)

I need to build a short example (please give me some time)…

However: My Connect6ion looks like this:

￼

So you might adjust the protocol to igniteworks

Best

Matthias

**Reply by fedyanabramovich** [reply]
WOW! Changed protocol to `igniteworks` and it realy works! Thank you so much!

**Comment by matsvitt** [comment]
Glad to hear that :)

**Reply by lomumve** [reply]
hello, I add requirements-local.txt under docker folder and add sqlalchemy-ignitex==0.5.6 in this file
after this. I run docker compose down -v and then docker compose -f docker-compose-non-dev.yml up
but when I try to connect with igniteworks://localhost:10800, it shows error 
 2025-02-14 02:50:43,058:ERROR:pyignite.connection:Failed to perform handshake, connection to node(address=localhost, port=10800) with protocol context ProtocolContext(version=(1, 7, 0), features=4) failed: [Errno 111] Connection refused
superset_app          | Traceback (most recent call last):
superset_app          |   File "/app/.venv/lib/python3.11/site-packages/pyignite/connection/connection.py", line 238, in connect
superset_app          |     result = self._connect_version()
superset_app          |              ^^^^^^^^^^^^^^^^^^^^^^^
superset_app          |   File "/app/.venv/lib/python3.11/site-packages/pyignite/connection/connection.py", line 268, in _connect_version
superset_app          |     self._socket.connect((self.host, self.port))
and 
![image](https://github.com/user-attachments/assets/5628b2a1-497c-4d38-a5c9-9e104a362fec)

Do you know if there is any possible reason for this. Thanks a lot

**Comment by matsvitt** [comment]
Hi there,


I assume you have an Apache Ignite instance running on port 10800 on localhost?

Best

Matthias

**Comment by lomumve** [comment]
Hi Matsvitt,
Yes, My workflows are:
1. Download Supperset code from github
2. Add requirements-local.txt under docker folder and add sqlalchemy-ignitex==0.5.6 in this file
3. Run Docker Compose down -v
4. Run Docker Compose -f docker-compose-non-dev.yml up
5. Forward remote ignite db to my local env with command kubectl --kubeconfig dev.yaml -n namespace -dev port-forward service/grid-web 10800:10800
6. And then try to connect in Supperset under Other type with connection as igniteworks://localhost:10800

I tested with below python code could connect success:

from sqlalchemy import create_engine

engine = create_engine("igniteworks://localhost:10800")
connection = engine.connect()

**Comment by matsvitt** [comment]
are you able to connect to ignite using a python script running locally?

**Reply by lomumve** [reply]
Hi matvitt,
yes, I could connect success with below two ways:
1. connect with pygridgain
from pygridgain import Client
client = Client()
client.connect("localhost", 10800)
2. connect with sqlalchemy
from sqlalchemy import create_engine
engine = create_engine("igniteworks://localhost:10800")
connection = engine.connect()

but will got error when try to connect in superset UI with connectionstring "igniteworks://localhost:10800"
![image](https://github.com/user-attachments/assets/47109e6d-3258-42ca-a9a0-f4a000fe8258)

**Comment by matsvitt** [comment]
Well, you are running Apache Ignite in its own Container, don’t you?
Hence from the “Superset” Container, “igniteworks://localhost:10800” is pointing to an Apache Ignite service running within the same (i.e. Superset) container -  which it (presumably) isn’t.

If you have the two containers running in the same docker subnet, then use the “name” of the Apache Ignite Container instead of localhost.

**Reply by lomumve** [reply]
hello matsvitt, 
Thanks for the reply , agree with you that the issue may caused for the Apache Ignite in running on my laptop directly and supperset is running inside the docker.

**Comment by rusackas** [comment]
Hey @matsvitt  - thanks for this info! If you'd be willing to add Ignite to[ this docs page](https://superset.apache.org/docs/configuration/databases), we'd really appreciate the contribution!

**Comment by matsvitt** [comment]
Hey,

Sure - do you need any input from me? Could I do it myself?

Best

Mats

**Reply by nmhungk52** [reply]
Hi @matsvitt 
I’m currently testing the integration of Superset with Apache Ignite (2.16) using the sqlalchemy-ignitex library. Overall, it works reasonably well; however, I encountered an issue related to datetime fields in the datetime dimension.

Superset is detecting the datetime fields as STRING in the dataset, which causes time-series analysis features to not work properly.

Do you have any suggestions or advice on how to resolve this issue?

**Comment by matsvitt** [comment]
Hi there,

Uhhh…that doesn’t sound good. It might be the case that the mapping is somehow insufficient. I would need to check into this -  However I am currently buried under work.

Maybe you have a look into

https://github.com/matsvitt/sqlalchemy-ignitex/blob/main/src/igniteworks/sqlalchemy/dialect.py

One would need to setup a whole test for this issue -  Unfortunately I have no time for this currently. 

Best

Matthias

**Reply by nmhungk52** [reply]
Thanks @matsvitt 
I’ve tried modifying the datat type mapping logic in dialect.py, but the results are still not very satisfactory. I’ll continue researching and working on this issue.
If you have time, please keep an eye on this issue together with me.

**Reply by lomumve** [reply]
hello @nmhungk52 , have you found a better way to fix the datetime fields query issue?
